### DISCUSSION TITLE
help! Ask a question

### DISCUSSION THREAD

**Discussion by yangleics3333** [marksgraham/OCT-Converter#114] [General]
Hi!
After installing OCT-Converter, I ran the e2e sample program and the program reported an error as follows. What is the problem??
I hope you can help me, I will not be grateful！
python version：3.10.9
numpy version：1.24.3 or 1.21.2 

Traceback (most recent call last):
  File "/Users/leiyang/Desktop/CODE/spruce-data-dp/patch_eye.py", line 1, in <module>
    from oct_converter.readers import E2E
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/oct_converter/readers/__init__.py", line 1, in <module>
    from .boct import BOCT
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/oct_converter/readers/boct.py", line 13, in <module>
    from oct_converter.image_types import OCTVolumeWithMetaData
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/oct_converter/image_types/__init__.py", line 2, in <module>
    from .oct import OCTVolumeWithMetaData
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/oct_converter/image_types/oct.py", line 6, in <module>
    import matplotlib.pyplot as plt
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/matplotlib/__init__.py", line 109, in <module>
    from . import _api, _version, cbook, docstring, rcsetup
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/matplotlib/rcsetup.py", line 28, in <module>
    from matplotlib.fontconfig_pattern import parse_fontconfig_pattern
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/matplotlib/fontconfig_pattern.py", line 15, in <module>
    from pyparsing import (Literal, ZeroOrMore, Optional, Regex, StringEnd,
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/pyparsing/__init__.py", line 144, in <module>
    from .helpers import *
  File "/Users/leiyang/anaconda3/lib/python3.10/site-packages/pyparsing/helpers.py", line 2, in <module>
    import html.entities
ModuleNotFoundError: No module named 'html.entities'

**Comment by yangleics3333** [comment]
<img width="1046" alt="image" src="https://user-images.githubusercontent.com/30650469/236204493-266514dc-4d57-4799-ae99-02f595a310b6.png">

**Comment by marksgraham** [comment]
Hi,

This seems to be some sort of package/dependency problem. I would suggest first starting a totally fresh anaconda environment, installing oct-converter in that, and seeing if that fixes things. If that doesn't work, it might also be worth trying with a different python version >=3.7 that is not 3.10

**Reply by yangleics3333** [reply]
First of all, thank you for your reply. The problem I mentioned above has been resolved. Thank you again.



Currently, I am running the sample program for e2e and the following code is reported. Could you please help me check what the problem is.

Thank you very much, and I am also looking forward to the moment when this program successfully runs.

The error message is as follows

**Comment by marksgraham** [comment]
I'm afraid I've seen this error once before and not been able to solve it. One suggestion is so try the eyepy readere [here](https://github.com/MedVisBonn/eyepy) - this can also read .e2e and is implemented a bit differently, so it is possible it will have more luck

**Reply by Oli4** [reply]
I tried to download the file, but I seem to be required to login for downloading the file. I wanted to try it with eyepy.

**Comment by marksgraham** [comment]
Hey,

I think I have fixed. Can you reinstall the main branch using
`pip install git+https://github.com/marksgraham/OCT-Converter`
and let me know if it works

**Reply by yangleics3333** [reply]
Hi Mark
Thank you for your response to this question. I haven't been able to reply to your message in the past few days because I have had health issues and have been recuperating for the past few days. I have seen your reply today and will test the new program tomorrow. I will provide you with the specific results as soon as possible. Thank you, Mark!

**Reply by yangleics3333** [reply]
Hi Mark
I have run the project you specified and the result is that the image can be successfully read and retrieved. But at the same time, there is also a bad news that the obtained data content is inconsistent with what appears on the device. I will now use screenshots to place two different data results below.

data ： e2e ;    SD-OCT
**oct-converter**

**Comment by marksgraham** [comment]
Hi,

Interesting. As a sanity check, does the data that is extracted from OCT-Converter also appear on the device? Or is is totally different?

**Comment by marksgraham** [comment]
Ok I'll take a look - can you just confirm how many fundus images and how many b-scans you expect in the .e2e file you sent me?

**Reply by yangleics3333** [reply]
Let me send the data to your email

**Reply by marksgraham** [reply]
I find 97 OCT b-scans and 1 fundus image when I run it through OCT-converter. What additional data do you expect it to extract that it doesn't?

**Reply by yangleics3333** [reply]
what！！？？
Why is that? Can I see your running program and the data results???
After running the e2e example program in oct-converter, I only obtained one image.

**Reply by marksgraham** [reply]
did you install with `pip install git+https://github.com/marksgraham/OCT-Converter` to get the version with the updated reader? if you justinstall with `pip install oct-converter` you wouldn't get the most up-to-date version

**Reply by marksgraham** [reply]
i just ran the simple example with the updated code:

```python
from oct_converter.readers import E2E
filepath = "/Users/mark/Downloads/data.e2e"
file = E2E(filepath)
oct_volumes =  file.read_oct_volume()

for volume in oct_volumes:
    #volume.peek(show_contours=True)  # plots a montage of the volume
    volume.save("{}_{}.png".format(volume.volume_id, volume.laterality))

fundus_images = (
    file.read_fundus_image()
)  # returns a list of all fundus images with additional metadata if available
for image in fundus_images:
    #plt.imshow(image.image)
    #plt.show()
    image.save("{}+{}.png".format(image.image_id, image.laterality))
```

**Reply by yangleics3333** [reply]
Thank you, Mark
I can also generate a complete dataset normally. Thank you for your help.
At the same time, I also want to ask if there is patient and other related information in E2E format data, and how to obtain this data?
For example, Dicom data, a Dicom file contains the image data itself and its related label information, such as Study ID, series ID, etc. Does E2E also have similar label data? How can I obtain them?

**Reply by marksgraham** [reply]
I've done a new release so you can get the updated version just running `pip install oct-converter`

**Comment by marksgraham** [comment]
Some metadata is extracted, you can see the options here:
https://github.com/marksgraham/OCT-Converter/blob/main/oct_converter/image_types/oct.py#L45-L55

and access with eg oct_volume.sex, oct_volume.laterality

**Reply by yangleics3333** [reply]
Hi Mark
Thank you for your technical support these days. I will continue to explore eye related imaging data and use this project. If there are any other issues, I will provide real-time feedback. Thank you in writing, Mark
